man/hfst-reweight.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-REWEIGHT "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-reweight \- =Reweight transducer weights simply
.SH SYNOPSIS
.B hfst-reweight
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Reweight transducer weights simply
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "Reweighting options:"
.TP
\fB\-a\fR, \fB\-\-addition\fR=\fI\,AVAL\/\fR
add AVAL to matching weights
.TP
\fB\-b\fR, \fB\-\-multiplier\fR=\fI\,BVAL\/\fR
multiply matching weights by BVAL
.TP
\fB\-F\fR, \fB\-\-function\fR=\fI\,FNAME\/\fR
operate matching weights by FNAME
.TP
\fB\-l\fR, \fB\-\-lower\-bound\fR=\fI\,LVAL\/\fR
match weights greater than LVAL
.TP
\fB\-u\fR, \fB\-\-upper\-bound\fR=\fI\,UVAL\/\fR
match weights less than UVAL
.TP
\fB\-I\fR, \fB\-\-input\-symbol\fR=\fI\,ISYM\/\fR
match arcs with input symbol ISYM
.TP
\fB\-O\fR, \fB\-\-output\-symbol\fR=\fI\,OSYM\/\fR
match arcs with output symbol OSYM
.TP
\fB\-S\fR, \fB\-\-symbol\fR=\fI\,SYM\/\fR
match arcs with input or output symbol SYM or both
.TP
\fB\-e\fR, \fB\-\-end\-states\-only\fR
match end states only, no arcs
.TP
\fB\-A\fR, \fB\-\-arcs\-only\fR
match arcs only, no end states
.TP
\fB\-T\fR, \fB\-\-tsv\-file\fR=\fI\,TFILE\/\fR
read reweighting rules from TFILE
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Format of result depends on format of INFILE
If AVAL, BVAL or FNAME are omitted, they default to neutral elements of addition, multiplication or identity function.
If LVAL or UVAL are omitted, they default to minimum and maximum values of the weight structure.
If ISYM, OSYM or SYM are omitted, they default to a value that matches all arcs.
Only one ISYM, OSYM and SYM can be given.
.PP
Float values are parsed with strtod(3) and integers strtoul(3).
The functions allowed for FNAME are <cmath> float functions with parameter count of 1 and a matching return value:
abs, acos, asin, ... sqrt, tan, tanh
.PP
The precedence of operands follows the formula BVAL * FNAME(w) + AVAL.
The formula is applied iff:
((LVAL <= w) && (w <= UVAL)),
where w is weight of arc, and
(ISYM == i) && (OSYM == o) && ((SYM == i) || (SYM == o)) ^^
(end state && \fB\-e\fR).
.PP
TFILE should contain lines with tab\-separated pairs of SYM and AVAL or BVAL. AVAL values must be preceded by a + character, BVAL should be given as plain digits. Comment lines starting with # and empty lines are ignored.
.PP
Weights are by default modified for all arcs and end states,
unless option \fB\-\-end\-states\-only\fR or \fB\-\-arcs\-only\fR is used.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-reweight home page:
<https://github.com/hfst/hfst/wiki/HfstReweight>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-reweight.cc
//! @file hfst-reweight.cc
//!
//! @brief Transducer reweighting tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif


#include <iostream>
#include <fstream>
#include <memory>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <cfloat>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "implementations/HfstBasicTransducer.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstOutputStream;
using hfst::implementations::HfstState;
using hfst::implementations::HfstBasicTransducer;
using hfst::implementations::HfstBasicTransition;


// add tools-specific variables here
static
float
id(float w)
{
  return w;
}
float addition = 0;
float multiplier = 1;
char* funcname = 0;
float (*func)(float) = id;
float upper_bound = FLT_MAX;
float lower_bound = 0;
char* input_symbol = 0;
char* output_symbol = 0;
char* symbol = 0;
bool ends_only = false;
bool arcs_only = false;
char* tsv_file_name = 0;
FILE* tsv_file = 0;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    // Usage line
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
           "Reweight transducer weights simply\n"
        "\n", program_name);

    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out, "Reweighting options:\n"
            "  -a, --addition=AVAL        add AVAL to matching weights\n"
            "  -b, --multiplier=BVAL      multiply matching weights by BVAL\n"
            "  -F, --function=FNAME       operate matching weights by FNAME\n"
            "  -l, --lower-bound=LVAL     match weights greater than LVAL\n"
            "  -u, --upper-bound=UVAL     match weights less than UVAL\n"
            "  -I, --input-symbol=ISYM    match arcs with input symbol ISYM\n"
            "  -O, --output-symbol=OSYM   match arcs with output symbol OSYM\n"
            "  -S, --symbol=SYM           match arcs with input or output symbol SYM or both\n"
            "  -e, --end-states-only      match end states only, no arcs\n"
            "  -A, --arcs-only            match arcs only, no end states\n"
            "  -T, --tsv-file=TFILE       read reweighting rules from TFILE\n"
            "\n");
    fprintf(message_out, "\n");
    print_common_unary_program_parameter_instructions(message_out);
    fprintf(message_out, "If AVAL, BVAL or FNAME are omitted, they default to neutral "
            "elements of addition, multiplication or identity function.\n"
            "If LVAL or UVAL are omitted, they default to minimum and maximum "
            "values of the weight structure.\n"
            "If ISYM, OSYM or SYM are omitted, they default to a value that "
            "matches all arcs.\nOnly one ISYM, OSYM and SYM can be given.\n\n"
            "Float values are parsed with strtod(3) and integers strtoul(3).\n"
            "The functions allowed for FNAME are <cmath> float functions with "
            "parameter count of 1 and a matching return value:\n"
            "abs, acos, asin, ... sqrt, tan, tanh\n\n"
            "The precedence of operands follows the formula "
            "BVAL * FNAME(w) + AVAL.\n"
            "The formula is applied iff:\n"
            "((LVAL <= w) && (w <= UVAL)),\n"
            "where w is weight of arc, and \n"
            "(ISYM == i) && (OSYM == o) && ((SYM == i) || (SYM == o)) ^^ \n"
            "(end state && -e).\n\n"
            "TFILE should contain lines with tab-separated pairs of SYM and "
            "AVAL or BVAL. AVAL values must be preceded by a + character, "
            "BVAL should be given as plain digits. "
            "Comment lines starting with # and empty lines are ignored.\n\n"
            "Weights are by default modified for all arcs and end states,\n"
            "unless option --end-states-only or --arcs-only is used.\n");
    fprintf(message_out, "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}


int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          HFST_GETOPT_UNARY_LONG,
          // add tool-specific options here
            {"addition", required_argument, 0, 'a'},
            {"multiplier", required_argument, 0, 'b'},
            {"function", required_argument, 0, 'F'},
            {"lower-bound", required_argument, 0, 'l'},
            {"upper-bound", required_argument, 0, 'u'},
            {"input-symbol", required_argument, 0, 'I'},
            {"output-symbol", required_argument, 0, 'O'},
            {"symbol", required_argument, 0, 'S'},
            {"end-states-only", no_argument, 0, 'e'},
            {"arcs-only", no_argument, 0, 'A'},
            {"tsv", required_argument, 0, 'T'},
            {0,0,0,0}
        };
        int option_index = 0;
        // add tool-specific options here
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "a:b:F:l:u:I:O:S:eT:A",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'a':
          addition = hfst_strtoweight(optarg);
          break;
        case 'b':
          multiplier = hfst_strtoweight(optarg);
          break;
        case 'F':
          funcname = hfst_strdup(optarg);
          if (strcmp(optarg, "cos") == 0)
            {
              func = &cosf;
            }
          else if (strcmp(optarg, "sin") == 0)
            {
              func = &sinf;
            }
          else if (strcmp(optarg, "tan") == 0)
            {
              func = &tanf;
            }
          else if (strcmp(optarg, "acos") == 0)
            {
              func = &acosf;
            }
          else if (strcmp(optarg, "asin") == 0)
            {
              func = &asinf;
            }
          else if (strcmp(optarg, "atan") == 0)
            {
              func = &atanf;
            }
          else if (strcmp(optarg, "cosh") == 0)
            {
              func = &coshf;
            }
          else if (strcmp(optarg, "sinh") == 0)
            {
              func = &sinhf;
            }
          else if (strcmp(optarg, "tanh") == 0)
            {
              func = &tanhf;
            }
          else if (strcmp(optarg, "exp") == 0)
            {
              func = &expf;
            }
          else if (strcmp(optarg, "log") == 0)
            {
              func = &logf;
            }
          else if (strcmp(optarg, "log10") == 0)
            {
              func = &log10f;
            }
          else if (strcmp(optarg, "sqrt") == 0)
            {
              func = &sqrtf;
            }
          else if (strcmp(optarg, "floor") == 0)
            {
              func = &floorf;
            }
          else if (strcmp(optarg, "ceil") == 0)
            {
              func = &ceilf;
            }
          else
            {
              error(EXIT_FAILURE, 0, "Cannot parse %s as function name",
                    optarg);
              return EXIT_FAILURE;
            }
          break;
        case 'l':
          lower_bound = hfst_strtoweight(optarg);
          break;
        case 'u':
          upper_bound = hfst_strtoweight(optarg);
          break;
        case 'I':
          input_symbol = hfst_strdup(optarg);
          break;
        case 'O':
          output_symbol = hfst_strdup(optarg);
          break;
        case 'S':
          symbol = hfst_strdup(optarg);
          break;
        case 'e':
          ends_only = true;
          break;
        case 'A':
          arcs_only = true;
          break;
        case 'T':
          tsv_file_name = hfst_strdup(optarg);
          break;
#include "inc/getopt-cases-error.h"
        }
    }

    if (arcs_only && ends_only)
      {
        error(EXIT_FAILURE, 0, "Options '--arcs-only' and '--end-states-only' cannot be used at the same time");
        return EXIT_FAILURE;
      }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    if (funcname == 0)
      {
        funcname = hfst_strdup("id");
      }
    if (upper_bound < lower_bound)
      {
        warning(0, 0, "Lower bound %f exceeds upper bound %f so reweight will"
                " never apply", lower_bound, upper_bound);
      }
    if (tsv_file_name != 0)
      {
        tsv_file = hfst_fopen(tsv_file_name, "r");
      }
    return EXIT_CONTINUE;
}

static
float
reweight(float w, const char* i, const char* o)
{
  if ((w < lower_bound) || (w > upper_bound))
    {
      // not within weight bounds, don't apply
      return w;
    }
  if ((i == 0) && (o == 0))
    {
      if (arcs_only)
        {
          return w;
        }
    }
  else if ((i != 0) && (o != 0))
    {
      if (ends_only)
        {
          return w;
        }
      if ((symbol != 0) && ((strcmp(i, symbol) != 0) &&
                            (strcmp(o, symbol) != 0) ) )
        {
          // symbol doesn't match, don't apply
          return w;
        }
      if ((input_symbol != 0) && (output_symbol != 0) &&
          (strcmp(i, input_symbol) != 0) && (strcmp(o, output_symbol) != 0))
        {
          // input doesn't match, don't apply
          return w;
        }
      else if ((input_symbol != 0) && (strcmp(i, input_symbol) != 0))
        {
          // input doesn't match, don't apply
          return w;
        }
      else if ((output_symbol != 0) && (strcmp(o, output_symbol) != 0))
        {
          // output doesn't match, don't apply
          return w;
        }
    }
  return multiplier * (*func)(w) + addition;

}

static
HfstTransducer&
do_reweight(HfstTransducer& trans)
  {
        HfstBasicTransducer original(trans);
        HfstBasicTransducer replication;
        HfstState state_count = 1;
        std::map<HfstState,HfstState> rebuilt;
        rebuilt[0] = 0;  // HfstBasicTransducer initially has state number zero
	if (original.is_final_state(0))
	  {
	    float nuweight = reweight(original.get_final_weight(0), 0, 0);
	    replication.set_final_weight(0, nuweight);
	  }
        HfstState source_state=0;
        for (HfstBasicTransducer::const_iterator state = original.begin();
             state != original.end();
             ++state)
          {
            if (rebuilt.find(source_state) == rebuilt.end())
              {
                replication.add_state(state_count);
                if (original.is_final_state(source_state))
                  {
		    float nuweight = reweight(original.get_final_weight(source_state), 0, 0);
                    replication.set_final_weight(state_count, nuweight);
                  }
                rebuilt[source_state] = state_count;
                state_count++;
              }
            for (const auto & arc : *state)
              {
                if (rebuilt.find(arc.get_target_state()) == rebuilt.end())
                  {
                    replication.add_state(state_count);
                    if (original.is_final_state(arc.get_target_state()))
                      {
			float nuweight = reweight(original.get_final_weight(arc.get_target_state()), 0, 0);
			replication.set_final_weight(state_count, nuweight);
                      }
                    rebuilt[arc.get_target_state()] = state_count;
                    state_count++;
                  }
                HfstBasicTransition nu(rebuilt[arc.get_target_state()],
                                       arc.get_input_symbol(),
                                       arc.get_output_symbol(),
                                       reweight(arc.get_weight(),
                                                arc.get_input_symbol().c_str(),
                                                arc.get_output_symbol().c_str()));
                replication.add_transition(rebuilt[source_state], nu);
              }
            source_state++;
          }
        trans = HfstTransducer(replication, trans.get_type());
        return trans;
  }


int
process_stream(HfstInputStream& instream, HfstOutputStream& outstream)
  {
    size_t transducer_n=0;
    while(instream.is_good())
    {
        transducer_n++;
        HfstTransducer trans(instream);
        char* inputname = hfst_get_name(trans, inputfilename);
        if (transducer_n==1)
        {
          verbose_printf("Reweighting %s...\n", inputname);
        }
        else
        {
          verbose_printf("Reweighting %s..." SIZE_T_SPECIFIER "\n", inputname, transducer_n);
        }
        if (NULL == tsv_file)
          {
            trans = do_reweight(trans);
            hfst_set_name(trans, trans, "reweight");
            hfst_set_formula(trans, trans, "W");
          }
        else
          {
            rewind(tsv_file);
            free(symbol);
            addition = 0;
            multiplier = 1;
            char* line = NULL;
            size_t len = 0;
            size_t linen = 0;
            verbose_printf("Reading reweights from %s\n", tsv_file_name);
            while (hfst_getline(&line, &len, tsv_file) != -1)
              {
                linen++;
                if (*line == '\n')
                  {
                    continue;
                  }
                if (*line == '#')
                  {
                    continue;
                  }
                const char* tab = strstr(line, "\t");
                if (NULL == tab)
                  {
                    if (*line == '#')
                      {
                        continue;
                      }
                    else
                      {
                        error_at_line(EXIT_FAILURE, 0, tsv_file_name, linen,
                                      "at least one tab required per line");
                      }
                  }
                const char* endstr = tab + 1;
                while ((*endstr != '\0') && (*endstr != '\n'))
                  {
                    endstr++;
                  }
                symbol = hfst_strndup(line, tab - line);
                char* weightspec = hfst_strndup(tab + 1, endstr - tab - 1);
                if (*weightspec == '+')
                  {
                    addition = hfst_strtoweight(weightspec + 1);
                  }
                else
                  {
                    multiplier = hfst_strtoweight(weightspec);
                  }
                free(weightspec);
                verbose_printf("Modifying weights %f < w < %f as %f * %s(w) + %f for symbol %s\n",
                   lower_bound, upper_bound, multiplier, funcname, addition, symbol);
                trans = do_reweight(trans);
              } // getline
              free(line);
            hfst_set_name(trans, trans, "reweight");
            hfst_set_formula(trans, trans, "W");
          } // if tsv_file
        outstream << trans.remove_epsilons();
        free(inputname);
      } // foreach transducer
    instream.close();
    outstream.close();
    return EXIT_SUCCESS;
  }



int main( int argc, char **argv ) {
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
  _setmode(1, _O_BINARY);
#endif

    hfst_set_program_name(argv[0], "0.1",
                          "HfstReweight");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n",
        inputfilename, outfilename);
    verbose_printf("Modifying weights %f < w < %f as %f * %s(w) + %f\n",
                   lower_bound, upper_bound, multiplier, funcname, addition);
    if (symbol)
      {
        verbose_printf("only if arc has symbol %s\n", symbol);
      }
    if (input_symbol)
      {
        verbose_printf("only if input symbol is %s\n", input_symbol);
      }
    if (output_symbol)
      {
        verbose_printf("only if output symbol is %s\n", output_symbol);
      }
    if (ends_only)
      {
        verbose_printf("only on final weights, no arcs\n");
      }
    if (arcs_only)
      {
        verbose_printf("only on arc weights, no end states\n");
      }
    // here starts the buffer handling part
    std::unique_ptr<HfstInputStream> instream;
    try {
      instream.reset((inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream());
    } catch(const HfstException e)  {
        error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
              inputfilename);
        return EXIT_FAILURE;
    }
    auto outstream = (outfile != stdout) ?
        std::make_unique<HfstOutputStream>(outfilename, instream->get_type()) :
        std::make_unique<HfstOutputStream>(instream->get_type());
    
    if ( is_input_stream_in_ol_format(*instream, "hfst-reweight"))
      {
        return EXIT_FAILURE;
      }

    retval = process_stream(*instream, *outstream);
    free(inputfilename);
    free(outfilename);
    return retval;
}

